### DISCUSSION TITLE
v10.3.0

### DISCUSSION THREAD

**Discussion by cspotcode** [TypeStrong/ts-node#1507] [Releases]
Discussion thread for the v10.3.0 release.

[Release notes](https://github.com/TypeStrong/ts-node/releases/tag/v10.3.0)

**Comment by elliottregan** [comment]
It looks like all the other conversations I've been following on the bug introduced in this release have died. 
- https://github.com/just-jeb/angular-builders/issues/1046
- https://github.com/tschaub/karma-source-map-support/issues/30#issuecomment-953219812

The `source-map-support` redirect that was introduced in this release causes issues with `karma-typescript`. I'm sure there is a technical reason that supported the change, but it seems irresponsible to blindly redirect all other libraries from `source-map-support` to  `@cspotcode/source-map-support`. The change breaks `karma-typescript`, and likely other lesser-used libraries.

As it is, I am forced to version-lock all my TypeScript + Karma projects to `ts-node@10.2.1` as I'm not aware of any way to get `karma-typescript`to import the correct `source-map-support` library.

To me, the right solution is to be more conservative with that redirect, or don't redirect at all. If another library wants to use `@cspotcode/source-map-support`, they should be able to choose to update their packages to do so.


---
**Example of a failure:**
https://github.com/elliottregan/primeNG-test/runs/5242554662?check_suite_focus=true#step:6:31

**Reply by cspotcode** [reply]
The weird thing about this issue, is that no one has implemented #1533, which is a pretty simple way to avoid this.  This makes me suspect that this issue is relatively minor and only affects a few people who were able to avoid it in other ways.

I've written about why we did this in a few other places, so I'll copy-paste my most recent explanation which links to a bunch of relevant information:

"@cspotcode/source-map-support" redirects require() of "source-map-support" to "@cspotcode/source-map-support" (https://github.com/cspotcode/node-source-map-support/pull/23), as a drop-in replacement with additional bugfixes and features. (https://github.com/cspotcode/node-source-map-support/issues/24) This is necessary because "@cspotcode/source-map-support" correctly handles multiple copies of the library, which avoids broken stack traces, but "source-map-support" does not. (https://github.com/cspotcode/node-source-map-support/pull/28, https://github.com/evanw/node-source-map-support/pull/209)

**Reply by cspotcode** [reply]
@elliottregan pretty sure you can fix this by modifying these 2x lines in karma-source-map-support:

https://github.com/tschaub/karma-source-map-support/blob/bcd5296c71bc805de24a91d2da5dfbb7f189df5f/lib/index.js#L3
https://github.com/tschaub/karma-source-map-support/blob/bcd5296c71bc805de24a91d2da5dfbb7f189df5f/lib/index.js#L12

**Reply by cspotcode** [reply]
@elliottregan have you given the above idea any thought?  Is there another ticket where I can follow along and give suggestions to the karma-source-map-support maintainers?

**Reply by magyargergo** [reply]
I have seen the same error and came here from https://github.com/just-jeb/angular-builders/issues/1046. I would be really happy to see this going forward. If you need someone to submit a PR please let me know. I will need more context about it though. For the record, I have `v10.5` of `ts-node` and I'm trying to upgrade to ESM.

This is the stack all I get:
```bash
Chrome 99.0.4844.51 (Mac OS 10.15.7) ERROR
  Uncaught ReferenceError: sourceMapSupport is not defined
  at node_modules/karma-source-map-support/lib/client.js:1:1
  
  ReferenceError: sourceMapSupport is not defined
      at node_modules/karma-source-map-support/lib/client.js:1:1
```

**Reply by cspotcode** [reply]
@magyargergo try implementing this fix in karma-source-map-support: https://github.com/TypeStrong/ts-node/discussions/1507#discussioncomment-2237936

**Reply by magyargergo** [reply]
@cspotcode I don't have the full context of the fix. Could you please provide me some links?

**Reply by cspotcode** [reply]
The linked code does this:

```
const smsPath = path.dirname(require.resolve('source-map-support'));
createPattern(path.join(smsPath, 'browser-source-map-support.js'))
```

This should be simplified to the following:

```
createPattern(require.resolve('source-map-support/browser-source-map-support.js'))
```

The flaw in the original code, is that it attempts to resolve `source-map-support` first.  The problem is that's the node version of source-map-support, but that's not actually what it wants: it wants the browser version.  It should attempt to resolve the browser version directly.

**Reply by magyargergo** [reply]
I tried it as you suggested but It didn't work. I modified the `node_modules/karma-source-map-support/lib/index.js` file to see if it fixes.
```javascript
const path = require('path');

const createPattern = function(pattern) {
  return {pattern: pattern, included: true, served: true, watched: false};
};

const init = function(files) {
  files.unshift(createPattern(path.join(__dirname, 'client.js')));
  files.unshift(
    createPattern(require.resolve('source-map-support/browser-source-map-support.js'))
  );
};

init.$inject = ['config.files'];

module.exports = {
  'framework:source-map-support': ['factory', init]
};
```
I checked other occurrences as well. However, I don't think they are related.
- `node_modules/@angular-devkit/build-angular/src/webpack/plugins/karma/karma.js`
- `node_modules/@angular-builders/custom-webpack/node_modules/@angular-devkit/build-angular/src/webpack/plugins/karma/karma.js`

**Reply by cspotcode** [reply]
Same stack trace, or different stack trace?

**Reply by magyargergo** [reply]
Sorry, it was the same. I didn't get time to get the bottom of it. Once I finished with my task, I'll look into it.

**Reply by cspotcode** [reply]
If the error and stack was identical, that indicates to me that your changes were ignored.  Perhaps they were made in the wrong place, or perhaps a package manager overwrote them.

**Reply by cspotcode** [reply]
I tested the fix myself to prove that it works.

**Reply by cspotcode** [reply]
Any update on this?  The proposed fix is proven to work, so the next step is to send a pull request to `karma-source-map-support`.

**Reply by xfh** [reply]
I got the same problem as @magyargergo and managed to solve it with your instructions @cspotcode. Thanks!

It's however not enough to modify `node_modules/karma-source-map-support/lib/index.js` as mentioned above.
The following adjustment was needed as well:
```
--- node_modules/@angular-devkit/build-angular/src/webpack/plugins/karma/karma.js.original	2022-04-13 23:27:12.358000000 +0200
+++ node_modules/@angular-devkit/build-angular/src/webpack/plugins/karma/karma.js	2022-04-13 23:31:06.887262014 +0200
@@ -82,10 +82,9 @@
         // Code taken from https://github.com/tschaub/karma-source-map-support.
         // We can't use it directly because we need to add it conditionally in this file, and karma
         // frameworks cannot be added dynamically.
-        const smsPath = path.dirname(require.resolve('source-map-support'));
         const ksmsPath = path.dirname(require.resolve('karma-source-map-support'));
         addKarmaFiles(config.files, [
-            { pattern: path.join(smsPath, 'browser-source-map-support.js'), watched: false },
+            { pattern: require.resolve('source-map-support/browser-source-map-support.js'), watched: false },
             { pattern: path.join(ksmsPath, 'client.js'), watched: false },
         ], true);
     }
```

I don't know how to move forward. karma-source-map-support repo seems quite dead...

**Reply by cspotcode** [reply]
Have you sent them a pull request with the necessary changes?
